[Feature Request] [Draft] [UI/UX] Allow to turn dimension arrows on/off individually

CalligaroV opened the pull request:
Fixes #7573

This PR adds the properties `ArrowTypeStart`, `ArrowTypeEnd` and all the code/settings/ui needed to manage these new two properties.

Roy-043 commented:
Thank you very much for working on this. I have had a quick look at the code. Overall it looks OK. There are a couple of issue that need attention though.

1. There should be code to update objects created in older versions.
2. You have worked on the code that is now used to suppress one of the arrows of a radial dimension. This code should probably be removed. Instead one of the arrows can be set to "None".
3. Wires now have a boolean property to toggle the end arrow. This property should be removed.
4. I am not sure if an end arrow makes sense for a Draft Label. If the end arrow is a dot it could touch or overlap the text.

Side note: I see that you are using the QT editor for the UI files. I hate how it saves the cells in a grid in a random order. It makes the GitHub diff useless.

yorikvanhavre commented:
Great work! Globally agree with @Roy-043's observations. Besides that, wires being able to also display a start arrow is a nice thing. Not sure your code sets/updates the correct location of the arrows, though? That might need to be checked

CalligaroV commented:
@Roy-043 @yorikvanhavre many thanks for your appreciation and your feedback! :smile:

@Roy-043 about your suggestions, I'll reply both here and on the code lines:
1. As happened for previous modifications, also in this case I agree with you.
I was thinking about doing something similar to what is done by `PathArray.migrate_properties_0v19`.
I take this as reference simply because sometimes I see the warning messages it generates when I open old files but if you have better ideas please let me know!
2. Comment in code
3. Comment in code
4. Comment in code
5. I was thinking about that too but haven't made any modification about it simply because I don't remember it among the original requests but I think if could be a nice idea!
@yorikvanhavre comments on that?

About the UI files, I agree with you on this point too. How do you usually modify this kind of file?

@yorikvanhavre about your comments:
- Are you saying that it could be better to keep the property @Roy-043 suggests to remove on point 3?
- In my tests I found that sometimes the direction of the arrows is flipped but I don't remember I saw them in a wrong position. However I must admit that I haven't focused much on this kind of possible issues

CalligaroV commented:
> 1. Something like `PathArray.migrate_properties_0v19` would be fine.

Great, I'll add some code similar to that one then.

> 2. 'Comment in code': I do not see a new commit. But maybe I misunderstand.

I added some comments (GitHub comments) for example in `view_dimension.py` at line 214.
If you don't see them is it possible that this kind of comment is "hidden" as long as PR is marked as draft like this one?

I apologize for the stupid question, it's the first time I mark a PR as draft :sweat_smile: 

> 3. We can perhaps implement 2 arrow sizes later. We can wait and see if users actually have a need for this.

Well if it's not something that needs to be implemented with this PR then I agree with you.
Otherwise we could ask comments about that on #7573 conversation or in the forum post.

> Regarding the UI files: I do this manually. But I can't ask others to do this as well. It is not very convenient.

Yes, it's a little bit more complicated but if this helps the review here on GitHub I'd say that it's worth the effort :wink:
I'll modify the UI file and submit an updated and "cleaner" version with the next commit.

Roy-043 commented:
Even after logging in I do not see comments. See screenshot.
![git-comment-missing](https://github.com/FreeCAD/FreeCAD/assets/70520633/fec7450d-4d96-4c3f-9c90-e2a4d1c4eb0b)

CalligaroV commented:
Ok then probably it's because of the status of the PR.

I'll change it right now, hopefully this will allow you to see my comments.
Just for reference, this is what I see right now in the conversation about this PR
![immagine](https://github.com/FreeCAD/FreeCAD/assets/144716783/86184473-db23-4757-8f81-e76e3063724e)

chennes commented:
@CalligaroV the "Pending" tag on each of those comments means you have not yet submitted your review, so you are the only one who can see those comments. On GitHub a "Review" works differently than normal comments, it's designed to allow you to make a whole bunch of comments all at once, so that the PR author doesn't get eleventy billion notifications (and so you can edit them before they get submitted, if as you go through the code you need to change an earlier comment).

CalligaroV commented:
Thanks @chennes for the advice! I tried to change the comments from review to simple comments, hope that now they are visible for everyone.

Can someone please confirm that?

p.s.: Sorry for this lack of basic knowledge of GitHub from my side!

Roy-043 commented on src/Mod/Draft/draftviewproviders/view_dimension.py:
```diff
@@ -203,15 +203,25 @@ def set_graphics_properties(self, vobj, properties):
-            vobj.ArrowType = utils.ARROW_TYPES
-            vobj.ArrowType = utils.ARROW_TYPES[params.get_param("dimsymbol")]
+            vobj.ArrowTypeStart = utils.ARROW_TYPES
+            vobj.ArrowTypeStart = utils.ARROW_TYPES[params.get_param("dimsymbolstart")]
```
Yes that is what I mean.

CalligaroV replied:
I wasn't sure, i thought that there could be a more elegant way to do that but if for you it's ok to manage this point this way I'll submit the modifications with the next commit.

Roy-043 commented on src/Mod/Draft/draftviewproviders/view_label.py:
```diff
@@ -102,15 +102,25 @@ def set_graphics_properties(self, vobj, properties):
             vobj.ArrowSize = params.get_param("arrowsize")
 
-        if "ArrowType" not in properties:
+        if "ArrowTypeStart" not in properties:
```
Yes, that is what I mean.

CalligaroV replied:
Got it, I'll submit the modifications with the next commit.

Roy-043 commented on src/Mod/Draft/draftviewproviders/view_wire.py:
```diff
@@ -114,14 +125,18 @@ def updateData(self, obj, prop):
 
     def onChanged(self, vobj, prop):
-        if prop in ["EndArrow","ArrowSize","ArrowType","Visibility"]:
+        if prop in ["EndArrow","ArrowSize","ArrowTypeStart","ArrowTypeEnd","Visibility"]:
```
Yes, of course this property would need to be handled by the `migrate_properties_0v22` function.

CalligaroV replied:
Ok!

I think I'll implement this as for `migrate_properties_0v19`, therefor adding one method to handle the wires and another method to handle the dimensions.

I'm thinking that way simply because I see that now there are two `migrate_properties_0v19` methods, one in `patharray.py` and the other in `pointarray.py` but if you prefer I implement the new method in a different way just let me know!

FEA-eng commented:
> Can someone please confirm that?

Yes, the comments are visible for everyone.

yorikvanhavre commented:
This seems good to me, looking at the code (I still need to test).
Regarding backwards compatibility, I think:
- We add a block of code to convert the old properties to the new ones when opening the file, in set_graphics_properties()
- We delete the old properties
- We warn the user that this happened in the report view
And that's it. I think we don't maintain the old properties. The new file will display wrongly on older FreeCAD versions, but it is easily fixable by the use, so IMHO it's no show stopper. We should still test new files open on older FreeCAD without error, though.
What do you think @Roy-043 ?

Roy-043 commented:
This PR still needs some work, see my initial comments.

yorikvanhavre commented:
Ok for me :)

marcuspollio commented:
Hi @CalligaroV 

Could you please let us know the status of this PR and rebase with latest main if it needs testing.

@yorikvanhavre @Roy-043

Is this PR considered for inclusion in version 1.0 or is it for later ?

Thank you !

Roy-043 commented:
This can be included in V1.0. I don't mind finishing the PR if @CalligaroV does not have the time.

CalligaroV commented:
Hi @marcuspollio and @Roy-043!

Sorry for keeping this PR untouched for so long, as you may know in the last moths I focused more on Toponaming related modifications.

I want to close this PR ASAP, also because what's left to do doesn't look too much.

I'll work on this in the next days, hope I can complete it in short time!

maxwxyz commented:
@CalligaroV just a heads up for the upcoming feature freeze: 
If you want this draft PR to be considered for a version 1.0 release, it should be marked _ready for review_ by **Monday, 2024-06-03** at the latest. (further details in the 1.0 feature freeze [forum announcement](https://forum.freecad.org/viewtopic.php?p=759605#p759605)).

CalligaroV commented:
Thanks for the update @maxwxyz!

I updated the branch for this PR, now it is ready for review.
I'll convert back to draft status if the CI fails for any reason.

@Roy-043 @yorikvanhavre @marcuspollio, can you please give a look at the last modifications to see if there's something wrong/missing?

In the last commit I added also the possibility to set the individual sizes of the arrows but if you want I can revert and keep only the other commits.

TIA!

yorikvanhavre commented:
The code looks good, but I'd still like to give it a test spin. Unfortunately I'm buried under a mountain of work for the next two weeks or so, so if anyone else wants to test and merge before that, please do!

CalligaroV commented:
Thanks for looking at this @yorikvanhavre!

I'll wait for feedback from others then.

Roy-043 commented:
I encounter multiple bugs and issues. This needs more work. Several of my comments made in January have not yet been addressed.

1. The start and end arrows of a Draft_Wire both appear at the endpoint.
3. The size of both arrows is controlled by ArrowSizeStart.
4. The wire still has an EndArrow property that you have to set to True.
5. You can't specify a 'None' arrow. This results in a "Symbol not implemented. Using a default symbol." warning.
6. `params.get_param("dimsymbol")` fails because that preference is no longer defined.
7. Radial dimensions have 2 arrows, one of which is always invisible. Which does not make sense.

I am switching this to Draft for now.

maxwxyz commented:
@CalligaroV will you pick up this PR again?

CalligaroV commented:
@maxwxyz yes sure, but not immediately.

At least after the release of v1.0.0.

There are some point of @Roy-043's https://github.com/FreeCAD/FreeCAD/pull/11941#issuecomment-2154796580 I'd like to discuss/understand better before moving on.

But for sure this is still in my todo list.

luzpaz commented:
@CalligaroV any time to revisit this PR ?

CalligaroV commented:
@luzpaz yes, I'm going back to this PR in these days.

By now I simply rebased it on top of 7083362acb8529017b58ea03bf7416da0192b25b and fixed the merge conflicts.

In the next couple of days I'll push another commit with modifications to address most of https://github.com/FreeCAD/FreeCAD/pull/11941#issuecomment-2154796580

CalligaroV commented:
@Roy-043, I have some questions about you last https://github.com/FreeCAD/FreeCAD/pull/11941#issuecomment-2154796580:

> 1. The start and end arrows of a Draft_Wire both appear at the endpoint.

Yes: now I understand also @yorikvanhavre's https://github.com/FreeCAD/FreeCAD/pull/11941#issuecomment-1887331984 
> Besides that, wires being able to also display a start arrow is a nice thing.

In this case I should update the code of `updateData` and `attach` (class `ViewProviderWire`) to get both the coordinates of the end point and the start point, right?

> 3. The wire still has an EndArrow property that you have to set to True (see Jan. comment).

AFAICS If I remove the `EndArrow` property then the arrows/markers will always be visible if the wire itself is visible.
Is this ok or do you have some suggestions about how to handle the removal of this property?

> 4. Radial dimensions have 2 arrows, one of which is always invisible. Which does not make sense (see Jan. comment).

I think now I understood this point.
Do you think it could be enough moving
```python
startSymbol = utils.ARROW_TYPES.index("None")
if vobj.Object.Diameter or not self.is_linked_to_circle():
    startSymbol = utils.ARROW_TYPES.index(vobj.ArrowTypeStart)
``` 
from `draw_dim_arrows` to `set_graphics_properties` (obviously replacing `startSymbol` with ` vobj.ArrowTypeStart`)?

Roy-043 commented:
Sorry for the delay, I had to refresh my memory.

> In this case I should update the code of `updateData` and `attach` (class `ViewProviderWire`) to get both the coordinates of the end point and the start point, right?

Yes, but also the `onChanged` function.

> AFAICS If I remove the `EndArrow` property then the arrows/markers will always be visible if the wire itself is visible.
> Is this ok or do you have some suggestions about how to handle the removal of this property?

As mentioned before, you would have to migrate this property, not just delete it. `ArrowTypeStart` should be `"None"` after migration because wires currently do not support a start arrow. `ArrowTypeEnd` would depend on the old value of `EndArrow` and the old value of `ArrowType`. if `EndArrow` is `False`, `ArrowTypeEnd` should be `"None"`, else `ArrowTypeEnd` should receive the old `ArrowType` value.

If you want I can assist with the coding. Just let me know.

Roy-043 commented:
@CalligaroV I have taken the liberty of rebasing this branch. I really would like to see this PR merged in the current dev cycle, but there are still a few loose ends if I recall correctly. You have been away from FreeCAD for some time now, so I do not know if you are still interested in working on this, and have the time to do so. If not, then I can finish this PR instead. Please let me know. Thanks.
